pip_audit/_dependency_source/poetry.py

```diff
@@ -0,0 +1,64 @@
+"""
+Collect dependencies from `poetry.lock` files.
+"""
+from __future__ import annotations
+
+import logging
+import subprocess
+import sys
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Iterator
+
+import toml
+from packaging.version import InvalidVersion, Version
+
+from pip_audit._dependency_source import DependencySource
+from pip_audit._fix import ResolvedFixVersion
+from pip_audit._service import Dependency, ResolvedDependency, SkippedDependency
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class PoetrySource(DependencySource):
+    """
+    Dependency sourcing from `poetry.lock`.
+    """
+
+    path: Path
+
+    def collect(self) -> Iterator[Dependency]:
+        """
+        Collect all of the dependencies discovered by this `PoetrySource`.
+        """
+        with self.path.open("r") as stream:
+            packages = toml.load(stream)
+        for package in packages["package"]:
+            name = package["name"]
+            try:
+                version = Version(package["version"])
+            except InvalidVersion:  # pragma: no cover
+                skip_reason = (
+                    "Package has invalid version and could not be audited: "
+                    f"{name} ({package['version']})"
+                )
+                logger.debug(skip_reason)
+                yield SkippedDependency(name=name, skip_reason=skip_reason)
+            else:
+                yield ResolvedDependency(name=name, version=version)
+
+    def fix(self, fix_version: ResolvedFixVersion) -> None:
+        """
+        Fixes a dependency version for this `PoetrySource`.
+
+        Requires poetry to be installed in the same env.
+
+        Note that poetry ignores the version we want to update to,
+        and goes straight to the latest version allowed in metadata.
+        """
+        subprocess.run(
+            [sys.executable, "-m", "poetry", "update", "--lock", fix_version.dep.name],
+            cwd=self.path.parent,
+            stdout=subprocess.DEVNULL,
+        ).check_returncode()
```

Comment:
Unfortunately, I don't think this is an acceptable tradeoff: we want users of `pip-audit` to trust that `pip-audit --fix` actually does the fix upgrades that the vulnerability reports suggest, not that it bumps to whatever `poetry update` thinks the latest acceptable version is.

Similarly, implicitly requiring users to have `poetry` installed is probably not an acceptable requirement for `pip-audit` -- we're developing it so that it *can* be integrated into `pip` at a later date, so executing non-`pip` programs is probably a non-starter.

The way I see it, we have two options here:

* We can try and manually modify the lockfile ourselves. You mentioned a bunch of hashes; assuming they're distribution hashes of the kind that `requirements.txt` supports, we could retrieve them from PyPI. I'm not sure how difficult this would be, since I'm not too familiar with Poetry or its internals 😅 
* We could return an error when a user tries to use `--fix` with a `PoetrySource`. This would be unlike any of our other current packages, but IMO it's not the worst choice if there's no straightforward/reliable way to perform the upgrades ourselves.

cc @di for thoughts.

Reply:
I looked into this a little more, and it looks like the hashes in question are indeed the distribution hashes from PyPI. So we'd have to do some network interaction for Poetry fixes, but that's not a huge deal.

More generally, the hashed bits look like this:

```toml
[metadata]
lock-version = "1.1"
python-versions = "^3.7"
content-hash = "7db62de4ee4f2831829fc08b6a3d190f177c62d5af7676e20fb228bb58274850"

[metadata.files]
attrs = [
    {file = "attrs-22.1.0-py2.py3-none-any.whl", hash = "sha256:86efa402f67bf2df34f51a335487cf46b1ec130d02b8d39fd248abfd30da551c"},
    {file = "attrs-22.1.0.tar.gz", hash = "sha256:29adc2665447e5191d0e7c568fde78b21f9672d344281d0c6e1ab085429b22b6"},
]
```

So we have to update each hash under each package under `metadata.files`, as well as the `content-hash`...maybe? I don't know what that one represents.

Reply:
Oh, `content-hash` seems to be a SHA256 of the `pyproject.toml`. We shouldn't be modifying that, so I _think_ that can be left unchanged?

Reply:
Let's leave a proper implementation out of scope. For now, I'll go with the second option and raise NotImplementedError TBH, I personally don't see a use case for `--fix` at all. If dependencies are easy to keep up-to-date, just keep the up-to-date on a daily basis, with dependabot or similar. If there are hard to keep up-to-date, autoupgrading them isn't an option. Even if some people really do use `--fix`, definitely not a majority of users. So, I think that can wait a bit.

Reply:
> So we have to update each hash under each package under metadata.files, as well as the content-hash...maybe? I don't know what that one represents.

It's not only updating hashes of files but also updating the list of available distributions for the release. It stretches what you have to put into `ResolvedFixVersion` quite a bit. The PyPI JSON API provides that info, but then you'll need to go and request it. Currently, PoetrySource is quite simple and self-contained. Don't make another dephell ;)

Reply:
Removed the `fix` code, updated the PR description ✅